src/phyto_arm/src/network_data_capture.py

```
@@ -0,0 +1,753 @@
+#!/usr/bin/env python3
+from dataclasses import dataclass, field as dc_field
+import json
+import select
+import socket
+import time
+from typing import Any, Dict, List, Tuple
+
+import rospy
+from std_msgs.msg import Float64, String, Int64, Bool, Float64MultiArray, Int64MultiArray, Int8MultiArray
+from ds_core_msgs.msg import RawData
+
+
+# Type mapping for ROS message types
+ROS_TYPE_MAP = {
+    "float": Float64,
+    "str": String,
+    "int": Int64,
+    "bool": Bool,
+    "float[]": Float64MultiArray,
+    "int[]": Int64MultiArray,
+    "bool[]": Int8MultiArray,
+}
+
+VALID_CONNECTION_TYPES = ["udp", "tcp"]
+VALID_PARSING_STRATEGIES = ["json_dict", "json_array", "raw", "delimited"]
+
+def validate_config(config: dict) -> Tuple[bool, str]:
+    """Validate complete network_data_capture configuration
+
+    Args:
+        config: Dict containing complete node configuration
+
+    Returns:
+        tuple[bool, str]: (is_valid, error_message)
+    """
+    if "topics" not in config:
+        return False, "network_data_capture config must contain 'topics' dictionary"
```

These return values are just eventually passed back out to `rospy.logerr()`. Can we just use Exceptions?

---

The return values are rolled up into an error message that makes it easy for the user to see exactly which topic/subtopic/field caused the error. These values would be hidden in a stack trace

---

The passing around of string errors like this is my biggest concern with this PR. There is a more idiomatic way to do this:

```python
class ConfigurationError(Exception):
    pass

def validate_config():
    if "topics" not in config:
        raise ConfigurationError("network_data_capture config must contain 'topics' dictionary")

try:
    validate_config()
except ConfigurationError as e:
    rospy.logerr(e.message)
    exit(1)
```

---

I'm not totally convinced that exceptions are the best way to assemble validation output, but I can make this work

---

Exceptions are basically just a language feature for what you are doing "manually": stopping execution and bubbling up errors from wherever they happen in the call stack to the nearest error handler. The manual way can be prone to mistakes, for instance if you make a call and forget to check its return value, the error is just silently dropped.

Other nice features are that you can chain errors (`raise X from Y`) instead of mask them, and you can use the type system to designate different error handlers rather than string matching.

---

Alright I'll tackle this in the morrow

---

Okay see dc7893c094589085f67302a93379b3853975327c